### PR TITLE
AV JSON Importer and JSON key handler

### PR DESCRIPTION
**Pull request by ricdezen**
### Alpha Vantage Importer
Added an Importer for Alpha Vantage:
- Takes the datetime keys and inserts them in each atom
- Inserts the metadata in each atom
- Inserts the atom array in the database

### JSON key handler
Just a module containing a script that can apply recursively any kind of function to all the keys in a json dictionary, as deep as possible.
Also contains premade methods to rename keys and turn them into lower case.

### CODE REVIEW

File: importer/av_json_data_importer.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+                Expects a datetime key and a dict value.
+        '''
+        new_data = self._rename_keys(json_file_contents)
+        new_data = self._add_timestamp(new_data)
+        atoms = self._extract_atoms(new_data)
+        atoms = self.__add_metadata_to_atoms(new_data[METADATA_KEY], atoms)
+        atoms = self.__fix_atoms_datetime(atoms)
+        self.database.write(DatabaseData("atoms_b", atoms))
+
+    def _add_timestamp(self, data_dict: dict):
```
Comment (CremaLuca) [inline importer/av_json_data_importer.py]: Is this method private? Doesn't private methods require two underscores?

https://www.geeksforgeeks.org/private-methods-in-python/

File: importer/av_json_data_importer.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    def __fix_atoms_datetime(self, atoms: list):
+        for atom in atoms:
+            atom[DATETIME_KEY] = datetime.strptime(
+                atom[DATETIME_KEY], "%Y-%m-%d %H:%M:%S"
+            ).strftime("%Y-%m-%d %H:%M:%S.%f")[:-3]
+        return atoms
+
+    def __add_metadata_to_atoms(self, metadata: dict, atoms: list):
+        '''
+        Adds each metadata field to all dictionary elements of the given list
```
Comment (CremaLuca) [inline importer/av_json_data_importer.py]: A little confusing, try:
"Adds each metadata field to all ~dictionary~ atoms of the given list"

File: importer/av_json_data_importer.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        new_data = self._rename_keys(json_file_contents)
+        new_data = self._add_timestamp(new_data)
+        atoms = self._extract_atoms(new_data)
+        atoms = self.__add_metadata_to_atoms(new_data[METADATA_KEY], atoms)
+        atoms = self.__fix_atoms_datetime(atoms)
+        self.database.write(DatabaseData("atoms_b", atoms))
+
+    def _add_timestamp(self, data_dict: dict):
+        '''
+        Copies the timestamp key inside each atom.
```
Comment (CremaLuca) [inline importer/av_json_data_importer.py]:
```suggestion
        Copies the timestamp key inside each atom as a dictionary entry.
```

**Review by CremaLuca — CHANGES_REQUESTED** [review]
Could you use `json_key_handler` on YahooJSONDataImporter too please?

File: importer/av_json_data_importer.py (hunk 1 above)
Comment (ricdezen) [reply importer/av_json_data_importer.py]: My bad, not sure where but I somehow got the idea they needed only one, will fix it

**Comment by ricdezen** [comment]
> 
> 
> Could you use `json_key_handler` on YahooJSONDataImporter too please?

Omw
vegitorunning

**Review by CremaLuca — APPROVED** [review]
Seems all ok now. GJ
OK

**Pull request merged**